### PR TITLE
feat(schedule): support download full log for running task

### PR DESCRIPTION
**Pull request by ysjemmm**
What type of PR is this?
type feat

What this PR does / why we need it:
support download full log

Which issue(s) this PR fixes:
Fixes #

Special notes for your reviewer:
fix 43x branch could not package docker image pipeline

Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ysjemmm** [comment]
cc @JessieWuJiexi  PR is here now

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        this.requestDispatcher = requestDispatcher;
+        this.dispatchChecker = dispatchChecker;
+        this.taskFrameworkService = taskFrameworkService;
+        this.jobDispatchChecker = jobDispatchChecker;
+        this.taskExecutorClient = taskExecutorClient;
+        this.loggerProperty = loggerProperty;
+        this.cloudObjectStorageService = cloudObjectStorageService;
+    }
+
+    public String getLog(Long scheduleTaskId, OdcTaskLogLevel level) {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:94]: getLogContent

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    private final RequestDispatcher requestDispatcher;
+    private final TaskDispatchChecker dispatchChecker;
+    private final TaskFrameworkService taskFrameworkService;
+    private final JobDispatchChecker jobDispatchChecker;
+    private final TaskExecutorClient taskExecutorClient;
+
+    private final LoggerProperty loggerProperty;
+    private final CloudObjectStorageService cloudObjectStorageService;
+
+    public ScheduledTaskLoggerService(ScheduleTaskService scheduleTaskService,
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:74]: what is the reason use constructor here ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        this.loggerProperty = loggerProperty;
+        this.cloudObjectStorageService = cloudObjectStorageService;
+    }
+
+    public String getLog(Long scheduleTaskId, OdcTaskLogLevel level) {
+        try {
+            return LogUtils.getLatestLogContent(getLogFile(scheduleTaskId, level), loggerProperty.getMaxLines(),
+                    loggerProperty.getMaxSize());
+        } catch (Exception e) {
+            log.warn("{}, taskId={}", LogUtils.DEFAULT_LOG_CONTENT, scheduleTaskId, e);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:99]: who is expected to read this log line? 
if not for user, may not use same content as DEfAULT_LOG_CONTENT

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        }
+    }
+
+    public File downloadLog(Long scheduleTaskId, OdcTaskLogLevel level) {
+        return getLogFile(scheduleTaskId, level);
+    }
+
+    @SneakyThrows
+    public String getFullLogDownloadUrl(Long scheduleId, Long scheduleTaskId, OdcTaskLogLevel level) {
+        if (ObjectUtil.isNotNull(cloudObjectStorageService) && cloudObjectStorageService.supported()) {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:110]: better handle isNull first, code will be more beautiful

File: server/integration-test/src/test/java/com/oceanbase/odc/service/task/TestLogUtils.java [hunk 5]
```diff
@@ -43,7 +43,7 @@ public void test_readPartLogContent() {
             a.write("go".getBytes());
             a.write("\n".getBytes());
             String s = LogUtils.getLatestLogContent(fileName, 2L, 100L);
-            Assert.assertTrue(s.endsWith("java\ngo\n"));
+            Assert.assertTrue(s.startsWith("java\ngo\n") && !s.startsWith("test\n"));
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/integration-test/src/test/java/com/oceanbase/odc/service/task/TestLogUtils.java:46]: 👌

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/ScheduleController.java [hunk 6]
```diff
@@ -122,6 +127,24 @@ public SuccessResponse<String> getTaskLog(@PathVariable Long scheduleId, @PathVa
         return Responses.success(scheduleService.getLog(scheduleId, taskId, logType));
     }
 
+    @ApiOperation(value = "GetFullLogDownloadUrl", notes = "get full log download url")
+    @RequestMapping(value = "/schedules/{scheduleId:[\\d]+}/tasks/{taskId:[\\d]+}/log/getDownloadUrl",
+            method = RequestMethod.GET)
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/ScheduleController.java:132]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import cn.hutool.core.io.FileUtil;
+import lombok.Data;
+
+@ConfigurationProperties(prefix = "odc.schedule.log")
+@Component
+@RefreshScope
+@Data
+public class LoggerProperty {
+
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java:34]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    // This is the directory path for storing temporary log files from scheduled tasks pods
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java:35]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    private String tempLogDir;
+
+    private String directory = "./log";
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/LoggerProperty.java:38]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java (hunk 1 above)
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:94]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java (hunk 2 above)
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:74]: using constructor injection gives more flexibility in managing dependencies and avoids adding too many @Autowired annotations to your classes，it is functionally equivalent to @Autowired

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java (hunk 4 above)
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:110]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java (hunk 3 above)
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/schedule/ScheduledTaskLoggerService.java:99]: I have changed it

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceService.java [hunk 10]
```diff
@@ -199,43 +192,17 @@ public FlowInstanceDetailResp executeTask(@NotNull Long id) throws IOException {
...
-            return taskLoggerService.getLogByTaskFramework(level, taskEntity.getJobId());
-        }
-        if (!dispatchChecker.isTaskEntityOnThisMachine(taskEntity)) {
-            // The task is not executing on current machine, need to forward the request
-            ExecutorInfo executorInfo = JsonUtils.fromJson(taskEntity.getExecutor(), ExecutorInfo.class);
-            DispatchResponse response = requestDispatcher.forward(executorInfo.getHost(), executorInfo.getPort());
-            return response.getContentByType(new TypeReference<SuccessResponse<String>>() {}).getData();
-        }
-        return taskService.getLog(taskEntity.getCreatorId(), taskEntity.getId() + "", taskEntity.getTaskType(), level);
+    public String getLog(@NotNull Long flowInstanceId, OdcTaskLogLevel level) throws IOException {
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceService.java:195]: I have used original 43x branch function is flowTaskInstanceService.getLogDownloadableTaskEntity() to check permission

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            LoggerProperty loggerProperty) {
+        this.flowTaskInstanceService = flowTaskInstanceService;
+        this.requestDispatcher = requestDispatcher;
+        this.dispatchChecker = dispatchChecker;
+        this.taskService = taskService;
+        this.loggerProperty = loggerProperty;
+    }
+
+    @SneakyThrows
+    public String getLog(OdcTaskLogLevel level, Long flowInstanceId) {
```
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java:69]: I have used original 43x branch function is flowTaskInstanceService.getLogDownloadableTaskEntity() to check permission，so below also had checked permission

[the remark this answers is not in the document]
Comment (yizhouxw) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java]: > I have used original 43x branch function is flowTaskInstanceService.getLogDownloadableTaskEntity() to check permission，so below also had checked permission

then you should add @SkipAuthorize annotation here

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        return getLogFile(taskEntity.getCreatorId(), taskEntity.getId() + "", taskEntity.getTaskType(),
+                OdcTaskLogLevel.ALL);
+    }
+
+    public File getLogFile(Long userId, String flowInstanceId, TaskType type, OdcTaskLogLevel logLevel) {
+        String logFilePath = taskService.getLogFilePath(userId, flowInstanceId, type, logLevel);
+        try {
+            return taskService.getLogFile(logFilePath);
+        } catch (NotFoundException ex) {
+            log.warn(ErrorCodes.TaskLogNotFound.getEnglishMessage(new Object[] {"Id", flowInstanceId}));
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java:118]: log.warn("Task log file not found, flowInstanceId={}, logFilePath={}", flowInstanceId, logFilePath);

Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java:118]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java]: > > I have used original 43x branch function is flowTaskInstanceService.getLogDownloadableTaskEntity() to check permission，so below also had checked permission
> 
> then you should add `@SkipAuthorize("internal authenticated")` annotation here

`@SkipAuthorize`   seems to mean skip authentication. Does it feel like this annotation should be put on the `getLogWithoutPermission()`,  which `getLog()` needs to verify the permission

[the remark this answers is not in the document]
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceLoggerService.java]: > > I have used original 43x branch function is flowTaskInstanceService.getLogDownloadableTaskEntity() to check permission，so below also had checked permission
> 
> then you should add `@SkipAuthorize("internal authenticated")` annotation here

put the method of `getLogContentWithoutPermission`

**Review by yizhouxw — DISMISSED** [review]
lgtm

**Review by yizhouxw — DISMISSED** [review]

**Review by yhilmare — DISMISSED** [review]
lgtm

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**